### PR TITLE
ovl_En_Gm OK

### PR DESCRIPTION
**Pull request by shawlucas**
OK

### CODE REVIEW

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 1]
```diff
@@ -7,8 +14,27 @@
 void EnGm_Init(Actor* thisx, GlobalContext* globalCtx);
 void EnGm_Destroy(Actor* thisx, GlobalContext* globalCtx);
 void EnGm_Update(Actor* thisx, GlobalContext* globalCtx);
+void EnGm_DrawProc(Actor* thisx, GlobalContext* globalCtx);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:17]:
```suggestion
void EnGm_Draw(Actor* thisx, GlobalContext* globalCtx);
```
even if draw is set in code, like `BgHakaMegane_Draw`

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 2]
```diff
@@ -7,8 +14,27 @@
...
+void func_80A3DD7C(EnGm* this, GlobalContext* globalCtx);
+void EnGm_ProcessChoiceIndex(EnGm* this, GlobalContext* globalCtx);
+void func_80A3DF00(EnGm* this, GlobalContext* globalCtx);
+void func_80A3DF60(EnGm* this, GlobalContext* globalCtx);
+void EnGm_UpdateEye(EnGm* this);
+void EnGm_SetTextID(EnGm* this);
+void func_80A3E090(EnGm* this);
+
+s32 func_80A3D7C8(void);
+
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:33]:
```suggestion
void func_80A3DF60(EnGm* this, GlobalContext* globalCtx);

```
those prototypes aren't needed, the functions are defined before being used

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 3]
```diff
@@ -20,37 +46,279 @@ const ActorInit En_Gm_InitVars = {
...
+        this->actor.speedXZ = 0.0f;
+        this->actor.gravity = -1.0f;
+        this->actor.velocity.y = 0.0f;
+    }
+}
+
+void EnGm_UpdateEye(EnGm* this) {
+    if (this->eyeTimer != 0) {
+        this->eyeTimer--;
+    } else if ((u8)(this->eyeIndex += 1) >= 3) {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:126]: I was about to suggest replacing 3 with `ARRAY_COUNT(sEyeTextures)` but you moved the data inside the draw function so nope 😅 

you can replace `this->eyeIndex += 1` with `++this->eyeIndex` "increment and return"

is the u8 cast necessary?

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 4]
```diff
@@ -20,37 +46,279 @@ const ActorInit En_Gm_InitVars = {
...
+        this->eyeIndex = 0;
+        this->eyeTimer = Math_Rand_ZeroFloat(60.0f) + 20.0f;
+    }
+}
+
+void EnGm_SetTextID(EnGm* this) {
+    switch (func_80A3D7C8()) {
+        case 0:
+            if (gSaveContext.infTable[11] & 1) {
+                this->actor.textId = 0x304B; // "If you can wait five or six years, it should be ready. OK?"
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:136]: Fig [mentioned](https://github.com/zeldaret/oot/pull/380#discussion_r491707947) that we shouldn't include text excerpts anymore

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 5]
```diff
@@ -20,37 +46,279 @@ const ActorInit En_Gm_InitVars = {
...
+
+    dx = this->talkPos.x - player->actor.posRot.pos.x;
+    dz = this->talkPos.z - player->actor.posRot.pos.z;
+
+    if (Flags_GetSwitch(globalCtx, this->actor.params)) {
+        EnGm_SetTextID(this);
+        this->subActionFunc = func_80A3DC44;
+    } else if (func_8002F194(&this->actor, globalCtx)) {
+        this->subActionFunc = func_80A3DBF4;
+    } else if (this->collider.base.maskA & 2 || (SQ(dx) + SQ(dz)) < 10000.0f) {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:170]:
```suggestion
    } else if (this->collider.base.maskA & 2 || (SQ(dx) + SQ(dz)) < SQ(100.0f)) {
```

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 6]
```diff
@@ -20,37 +46,279 @@ const ActorInit En_Gm_InitVars = {
...
+    this->actor.posRot2.rot.z = this->actor.posRot.rot.z;
+    EnGm_UpdateEye(this);
+    SkelAnime_FrameUpdateMatrix(&this->skelAnime);
+    CollisionCheck_SetOC(globalCtx, &globalCtx->colChkCtx, &this->collider.base);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Gm/func_80A3DF60.s")
+void EnGm_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnGm* this = THIS;
+    this->actionFunc(this, globalCtx);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:278]:
```suggestion
    EnGm* this = THIS;
    
    this->actionFunc(this, globalCtx);
```

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 7]
```diff
@@ -20,37 +46,279 @@ const ActorInit En_Gm_InitVars = {
...
+    ICHAIN_F32(unk_4C, 30, ICHAIN_STOP),
+};
+
+void EnGm_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnGm* this = THIS;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    osSyncPrintf(VT_FGCOL(GREEN) "%s[%d] : 中ゴロン[%d]" VT_RST "\n", "../z_en_gm.c", 133,
+                 this->actor.params); // "%s[%d]: Medi Goron [%d]"
+    this->objBankIndex = Object_GetIndex(&globalCtx->objectCtx, OBJECT_GM);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:67]:
```suggestion
    this->objGmBankIndex = Object_GetIndex(&globalCtx->objectCtx, OBJECT_GM);
```

**Review by Dragorn421 — CHANGES_REQUESTED** [review]
I didn't know he had special text for showing him his (broken) sword, nice

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c (hunk 3 above)
Comment (shawlucas) [reply src/overlays/actors/ovl_En_Gm/z_en_gm.c:126]: Yeah the cast is necessary

**Comment by shawlucas** [comment]
Made suggested changes

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 8]
```diff
@@ -20,37 +41,280 @@ const ActorInit En_Gm_InitVars = {
...
+            case 2:
+                this->subActionFunc = EnGm_ProcessChoiceIndex;
+            default:
+                return;
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Gm/func_80A3D9D8.s")
+        this->subActionFunc = EnGm_ProcessChoiceIndex;
+    }
+    if (this->collider.base.maskA & 2 || (SQ(dx) + SQ(dz)) < 10000.0f) {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:205]:
```suggestion
    if (this->collider.base.maskA & 2 || (SQ(dx) + SQ(dz)) < SQ(100.0f)) {
```

File: src/overlays/actors/ovl_En_Gm/z_en_gm.h [hunk 9]
```diff
@@ -6,9 +6,21 @@
...
+    /* 0x014C */ SkelAnime skelAnime;
+    /* 0x0190 */ Vec3s limbDrawTable[18];
+    /* 0x01FC */ Vec3s transitionDrawTable[18];
+    /* 0x0268 */ EnGmActionFunc actionFunc;
+    /* 0x026C */ EnGmActionFunc subActionFunc;
+    /* 0x0270 */ ColliderCylinder collider;
+    /* 0x02BC */ s8 objGmBankIndex;
+    /* 0x02BE */ s16 counter;
+    /* 0x02C0 */ s16 eyeTimer;
+    /* 0x02C2 */ u8 eyeIndex;
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.h:22]: Wouldn't it be better to use plural for "eye" everywhere? Medi-goron has two eyes iirc

**Review by Dragorn421 — CHANGES_REQUESTED** [review]
very small things

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 10]
```diff
@@ -20,37 +41,280 @@ const ActorInit En_Gm_InitVars = {
...
+        this->actor.speedXZ = 0.0f;
+        this->actor.gravity = -1.0f;
+        this->actor.velocity.y = 0.0f;
+    }
+}
+
+void EnGm_UpdateEye(EnGm* this) {
+    if (this->eyeTimer != 0) {
+        this->eyeTimer--;
+    } else if ((u8)(++this->eyeIndex) >= 3) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:121]: Is the cast really needed here? `eyeIndex` is already a u8 so the cast is probably an optimization. You might have to not inline the ++ for it to work properly tho.
```suggestion
    } else {
        this->eyeIndex++;
        if (this->eyeIndex >= 3) {
```

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 11]
```diff
@@ -20,37 +41,280 @@ const ActorInit En_Gm_InitVars = {
...
+    }
+    if (this->collider.base.maskA & 2 || (SQ(dx) + SQ(dz)) < 10000.0f) {
+        this->collider.base.acFlags &= ~2;
+        func_8002F2CC(&this->actor, globalCtx, 415.0f);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Gm/func_80A3DA50.s")
+void func_80A3DD7C(EnGm* this, GlobalContext* globalCtx) {
+    u8 flag = func_8010BDBC(&globalCtx->msgCtx);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:212]: We usually call that `dialogState`, tho that will probably change later but it's better than `flag` for now imo
```suggestion
    u8 dialogState = func_8010BDBC(&globalCtx->msgCtx);
```

File: src/overlays/actors/ovl_En_Gm/z_en_gm.h [hunk 12]
```diff
@@ -6,9 +6,21 @@
...
+typedef void (*EnGmActionFunc)(struct EnGm*, GlobalContext*);
+
 typedef struct EnGm {
     /* 0x0000 */ Actor actor;
-    /* 0x014C */ char unk_14C[0x184];
+    /* 0x014C */ SkelAnime skelAnime;
+    /* 0x0190 */ Vec3s limbDrawTable[18];
+    /* 0x01FC */ Vec3s transitionDrawTable[18];
+    /* 0x0268 */ EnGmActionFunc actionFunc;
+    /* 0x026C */ EnGmActionFunc subActionFunc;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.h:17]: Based on the usage, my guess is these are supposed to be differentiated as update modes and actions. Probably good to hear the opinion of others on this case tho.
```suggestion
    /* 0x0268 */ EnGmUpdateFunc updateFunc;
    /* 0x026C */ EnGmActionFunc actionFunc;
```

File: src/overlays/actors/ovl_En_Gm/z_en_gm.c [hunk 13]
```diff
@@ -1,4 +1,11 @@
+/*
+ * File: z_en_gm.c
+ * Overlay: ovl_En_Gm
+ * Description: Medi-Goron
+ */
+
 #include "z_en_gm.h"
+#include <vt.h>
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Gm/z_en_gm.c:8]:
```suggestion
#include "vt.h"
```

**Review by fig02 — APPROVED** [review]
skelanime updated and review changes implemented

**Review by Roman971 — APPROVED** [review]

**Pull request merged**